src/example5.cpp

```diff
@@ -0,0 +1,264 @@
+            "        gl_Position = vec4(0.6, 0.6, 0.6, 1);\n"
+            "    else if (floating32_test > float(1))\n"
+            "        gl_Position = vec4(0.7, 0.7, 0.7, 1);\n"
+            "    else if (half_test > float(1))\n"
+            "        gl_Position = vec4(0.8, 0.8, 0.8, 1);\n"
+            "    else\n"
+            "        gl_Position = vec4(0);\n"
+            "}",
+
+            /* Fragment shader */
+            "#version 330\n"
+            "out vec4 color;\n"
+            "void main() {\n"
+            "    color = vec4(0.0, 0.0, 0.0, 1.0);\n"
+            "}"
+        );
+
+        mSerialShader.bind();
+        // upload uniforms, downloading is non-trivial and so is avoided.
+        // this tests the internal serialization structs just by compiling
+        mSerialShader.setUniform<uint32_t>("uint32_t_test", (uint32_t) 2);
+        mSerialShader.setUniform<int32_t>("int32_t_test",   (int32_t) -2);
+        mSerialShader.setUniform<uint16_t>("uint16_t_test", (uint16_t) 3);
+        mSerialShader.setUniform<int16_t>("int16_t_test",   (int16_t) -3);
+        mSerialShader.setUniform<uint8_t>("uint8_t_test",    (uint8_t) 4);
+        mSerialShader.setUniform<int8_t>("int8_t_test",      (int8_t) -4);
+        mSerialShader.setUniform<double>("floating64_test",   (double) 5);
+        mSerialShader.setUniform<float>("floating32_test",     (float) 6);
+        // vvv does not compile, incomplete type
+        // mSerialShader.setUniform<half_float::half>("half_test",  7);
+
+        this->performLayout();
+    }
+
+    ~ExtendedExampleApplication() { mSerialShader.free(); }
+
+protected:
+    /* Map of names to widgets.  We're going to save each widget in their own
+     * "namespace" in the serializer so that everything is saved rather than
+     * potentially overwriting some things.
+     */
+    std::map<std::string, nanogui::Widget*> mSerialMap;
+
+    /* This example will continuously overwrite the same file, in your own code
+     * you'll probably want to create new files or something if you want to
+     * be able to save and load multiple different snapshots.
+     */
+    std::string mSerialFilename = "nanogui_example_5_serialized";
+
+    /* Since we're using the same file each time, make sure that we've saved at
+     * least once before loading.
+     */
+    bool atLeastOneSave = false;
+
+    // Serialization shader test, does nothing
+    nanogui::GLShader mSerialShader;
+};
+
+// same as example1 loop, only uses `ExtendedExampleApplication`
+int main(void) {
+    try {
+        nanogui::init();
+
+        /* scoped variables */ {
+            nanogui::ref<ExtendedExampleApplication> app = new ExtendedExampleApplication();
+            app->drawAll();
+            app->setVisible(true);
+            nanogui::mainloop();
+        }
+
+        nanogui::shutdown();
+    }
+    catch (const std::runtime_error &e) {
+        std::string error_msg = std::string("Caught a fatal error: ") + std::string(e.what());
```

Comment:
To be honest, I don't even remember why I have this stuff in here.  Should the `try-catch` just be removed?